### PR TITLE
Add TypeScript types

### PR DESCRIPTION
**Pull request by zirkelc**
This PR adds TypeScript types as discussed in #370 

First, you need to use the official TypeScript [@serverless/typescript](https://github.com/serverless/typescript#readme) package. for your `serverless.ts` file.

TypeScript merges the types from this package with the base types from @serverless/typescript with either of these two options:

- Add `/// <reference types="serverless-step-functions" />` at the top of your `serverless.ts`
```ts
/// <reference types="serverless-step-functions" />
import type { AWS as Serverless } from '@serverless/typescript';
const serverlessConfiguration: Serverless = {
  // ...
};
```

- Add `"types": ["serverless-step-functions"]` to your `tsconfig.json`
```json5
{
  "compilerOptions": {
    // ...
    "types": [
      "serverless-step-functions"
    ]
  },
  "include": [
    "src/**/*.ts",
    "serverless.ts",
  ],
}
```

I can't guarantee that the types are fully complete and correct, but I think there are a good starting point.

If you rather prefer to have the types provided as a separate `@types/serverless-step-functions` package, I'd be willing to submit a PR to the [DefinetlyTypes](https://github.com/DefinitelyTyped/DefinitelyTyped) package.

### CODE REVIEW

**Comment by ebisbe** [comment]
Aren't `id` and `name` not mandatory?

File: lib/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+interface Fail extends StateBase {
+  Type: 'Fail';
+  Cause?: string;
+  Error?: string;
+}
+
+interface Map extends StateBase {
+  Type: 'Map';
+  ItemsPath: string;
+  Iterator: Definition;
```
Comment (ebisbe) [inline lib/index.d.ts:144]: Missing: 
```
ItemSelector?: {
    [key: string]: string | Array<unknown> | { [key: string]: string };
  }
```

File: lib/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+    stepFunctions?: {
+      stateMachines: StateMachines;
+      validate?: boolean;
+    };
+  }
+}
+
+type StateMachines = {
+  [stateMachine: string]: {
+    id: string;
```
Comment (ebisbe) [inline lib/index.d.ts:14]:
```suggestion
    id?: string;
```

File: lib/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+    name: string;
```
Comment (ebisbe) [inline lib/index.d.ts:15]:
```suggestion
    name?: string;
```

File: lib/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+};
+
+type TracingConfig = {
+  enabled: boolean;
+};
+
+type Definition = {
+  Comment?: string;
+  StartAt: string;
+  States: States;
```
Comment (ebisbe) [inline lib/index.d.ts:45]:
```suggestion
  States: States;
  ProcessorConfig?: {
    Mode: "INLINE" | "DISTRIBUTED";
  };
```

File: lib/index.d.ts [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+    };
+  }
+}
+
+type StateMachines = {
+  [stateMachine: string]: {
+    id: string;
+    name: string;
+    definition: Definition;
+    tracingConfig?: TracingConfig;
```
Comment (ebisbe) [inline lib/index.d.ts:17]:
```suggestion
    tracingConfig?: TracingConfig;
    events?: any;
    dependsOn?: any;
```

Not really sure what would be their types.

File: lib/index.d.ts (hunk 1 above)
Comment (ebisbe) [inline lib/index.d.ts:144]:
```suggestion
  Iterator: Definition;
  ItemSelector: {
    [key: string]: string;
  }
```

**Review by ebisbe — COMMENTED** [review]
That's what I've found missing with my current step function I'm building.

**Comment by ebisbe** [comment]
Also when I try to export the state machine to a single file it fails with:

```
Cannot load "serverless.ts": Initialization error: TSError: ⨯ Unable to compile TypeScript:
serverless.ts:650:7 - error TS2322: Type '{ events: { schedule: { rate: string; enabled: boolean; }; }[]; dependsOn: string[]; definition: { StartAt: string; States: { "Add empty LastEvaluatedKey": { Type: string; Next: string; Result: { LastEvaluatedKey: string; }; ResultPath: string; }; ... 4 more ...; "Add new LastEvaluatedKey": { ...; }; }; }; }' is not assignable to type '{ id?: string; name?: string; events?: any; dependsOn?: any; definition: Definition; tracingConfig?: TracingConfig; }'.
  The types of 'definition.States' are incompatible between these types.
    Type '{ "Add empty LastEvaluatedKey": { Type: string; Next: string; Result: { LastEvaluatedKey: string; }; ResultPath: string; }; Scan: { Type: string; Next: string; Parameters: { TableName: { Ref: string; }; FilterExpression: string; ExpressionAttributeNames: { ...; }; ExpressionAttributeValues: { ...; }; "ExclusiveStart...' is not assignable to type 'States'.
      Property '"Add empty LastEvaluatedKey"' is incompatible with index signature.
        Type '{ Type: string; Next: string; Result: { LastEvaluatedKey: string; }; ResultPath: string; }' is not assignable to type 'Choice | Fail | Map | Task | Parallel | Pass | Wait | Succeed'.
          Type '{ Type: string; Next: string; Result: { LastEvaluatedKey: string; }; ResultPath: string; }' is not assignable to type 'Pass'.
            Types of property 'Type' are incompatible.
              Type 'string' is not assignable to type '"Pass"'.

650       addCredits,
          ~~~~~~~~~~

  node_modules/serverless-step-functions/lib/index.d.ts:13:3
     13   [stateMachine: string]: {
          ~~~~~~~~~~~~~~~~~~~~~~~~~
     14     id?: string;
        ~~~~~~~~~~~~~~~~
    ... 
     19     tracingConfig?: TracingConfig;
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     20   };
        ~~~~
    The expected type comes from this index signature.
    ```

File: lib/index.d.ts (hunk 2 above)
Comment (ebisbe) [inline lib/index.d.ts:14]:
```suggestion
    id: string;
    type?: 'STANDARD' | 'EXPRESS'; 
```

File: lib/index.d.ts [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+  interface AWS {
+    stepFunctions?: {
+      stateMachines: StateMachines;
+      validate?: boolean;
+    };
+  }
+}
+
+type StateMachines = {
+  [stateMachine: string]: {
```
Comment (ebisbe) [inline lib/index.d.ts:13]:
```suggestion
  [stateMachine: string]: {
    type?: 'STANDARD' | 'EXPRESS'; 
```

File: lib/index.d.ts (hunk 5 above)
Comment (zirkelc) [reply lib/index.d.ts:17]: Looking at the source code:
https://github.com/serverless-operations/serverless-step-functions/blob/28fa07b254bd2f6fef934fe940911dd36fa77615/lib/deploy/stepFunctions/compileStateMachines.schema.js#L37-L40
https://github.com/serverless-operations/serverless-step-functions/blob/28fa07b254bd2f6fef934fe940911dd36fa77615/lib/deploy/stepFunctions/compileStateMachines.schema.js#L63

```suggestion
    tracingConfig?: TracingConfig;
    events?: any[];
    dependsOn?: string | string[];
```

**Comment by zirkelc** [comment]
> Also when I try to export the state machine to a single file it fails with:
> 

What do you mean with exporting it to a single file? The types use module augmentation to merge the types into the official AWS type:

https://github.com/serverless-operations/serverless-step-functions/blob/c464dcdc0e0255cd93170486c7fd907b9893c020/lib/index.d.ts#L3-L10

If you want to use the types as stand-alone, we should probably export the `stepFunctions` type separately.

Is that what you mean?

**Comment by ebisbe** [comment]
> > Also when I try to export the state machine to a single file it fails with:
> 
> What do you mean with exporting it to a single file? The types use module augmentation to merge the types into the official AWS type:
> 
> https://github.com/serverless-operations/serverless-step-functions/blob/c464dcdc0e0255cd93170486c7fd907b9893c020/lib/index.d.ts#L3-L10
> 
> If you want to use the types as stand-alone, we should probably export the `stepFunctions` type separately.
> 
> Is that what you mean?

Nope. What I wanted to have in a different file is the stateMachine implementation. Not the types. So I can have a file for each StateMachine and split things so I don't have a huge serverless.ts file. 

I've also found that we are missing `type?: 'STANDARD' | 'EXPRESS'; ` and `loggingConfig?: any;` ( https://github.com/serverless-operations/serverless-step-functions/issues/281 )

**Comment by zirkelc** [comment]
> > > Also when I try to export the state machine to a single file it fails with:
> > 
> > 
> > What do you mean with exporting it to a single file? The types use module augmentation to merge the types into the official AWS type:
> > https://github.com/serverless-operations/serverless-step-functions/blob/c464dcdc0e0255cd93170486c7fd907b9893c020/lib/index.d.ts#L3-L10
> > 
> > If you want to use the types as stand-alone, we should probably export the `stepFunctions` type separately.
> > Is that what you mean?
> 
> Nope. What I wanted to have in a different file is the stateMachine implementation. Not the types. So I can have a file for each StateMachine and split things so I don't have a huge serverless.ts file.
> 

Yes, I think the issue wouldn't appear if you could type your stateMachine like this:

```ts
export const stateMachine1: StateMachine = { 
  // ...
};
```

I assume the issue is related to [type widening](https://mariusschulz.com/blog/literal-type-widening-in-typescript) by TypeScript. So the last error message as an example:
> Types of property 'Type' are incompatible.
>  Type 'string' is not assignable to type '"Pass"'.

The state machine expects `type` to be a string literal of `Pass` or `Task` etc. But when you define your state machine outside of the serverless type, TypeScript widens the type from literal `Pass` to `string`. That causes the type error I guess.

**Comment by ebisbe** [comment]
> I assume the issue is related to [type widening](https://mariusschulz.com/blog/literal-type-widening-in-typescript) by TypeScript. So the last error message as an example:
> 
> > Types of property 'Type' are incompatible.
> > Type 'string' is not assignable to type '"Pass"'.
> 
> The state machine expects `type` to be a string literal of `Pass` or `Task` etc. But when you define your state machine outside of the serverless type, TypeScript widens the type from literal `Pass` to `string`. That causes the type error I guess.

Ok, I didn't know the term or what was really happening but that's exactly the issue. Then what we could do is export the single type `StateMachine`

File: lib/index.d.ts [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+    type?: 'STANDARD' | 'EXPRESS'; 
+    id?: string;
+    name?: string;
+    definition: Definition;
+    tracingConfig?: TracingConfig;
+  };
+};
```
[the remark this answers is not in the document]
Comment (zirkelc) [reply lib/index.d.ts]: Yes, that should work. Can you give it a try?

File: lib/index.d.ts [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+type StateMachines = {
+  [stateMachine: string]: StateMachine
+};
+
+export type StateMachine = {
+  type?: 'STANDARD' | 'EXPRESS';
+  id?: string;
+  name?: string;
+  definition: Definition;
+  tracingConfig?: TracingConfig;
```
Comment (ebisbe) [inline lib/index.d.ts:21]:
```suggestion
  tracingConfig?: TracingConfig;
  loggingConfig?: LoggingConfig;
  events?: any[];
  dependsOn?: string | string[];
};

type LoggingConfig = {
  level: "ERROR" | "ALL" | "FATAL" | "OFF";
  includeExecutionData: boolean;
  destinations: Resource | Resource[]
```
We forgot that one in one subcomment.

File: lib/index.d.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (ebisbe) [reply lib/index.d.ts]: That's working.

**Comment by ebisbe** [comment]
Everything is working from my side. Not sure if we are missing anything.

File: lib/index.d.ts [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+
+interface Fail extends StateBase {
+  Type: 'Fail';
+  Cause?: string;
+  Error?: string;
+}
+
+interface Map extends StateBase {
+  Type: 'Map';
+  ItemsPath: string;
```
Comment (ebisbe) [inline lib/index.d.ts:158]:
```suggestion
  ItemsPath?: string;
```

File: lib/index.d.ts [hunk 10]
```diff
@@ -0,0 +1,259 @@
...
+  Iterator: Definition;
+  ItemSelector: {
```
Comment (ebisbe) [inline lib/index.d.ts:160]:
```suggestion
  ItemSelector?: {
```

File: lib/index.d.ts [hunk 11]
```diff
@@ -0,0 +1,259 @@
+import { AWS } from '@serverless/typescript';
+
+declare module '@serverless/typescript' {
+  interface AWS {
+    stepFunctions?: {
+      stateMachines: StateMachines;
+      validate?: boolean;
```
Comment (ebisbe) [inline lib/index.d.ts:7]:
```suggestion
      validate?: boolean;
      noOutput?: boolean
```

File: lib/index.d.ts [hunk 12]
```diff
@@ -0,0 +1,260 @@
...
+  Next: string;
+};
+
+type ChoiceRuleOr = {
+  Or: ChoiceRuleComparison[];
+  Next: string;
+};
+
+type ChoiceRuleSimple = ChoiceRuleComparison & {
+  Next: string;
```
Comment (ebisbe) [inline lib/index.d.ts:140]:
```suggestion
  Next: string;
  Comment?: string:
```

File: lib/index.d.ts [hunk 13]
```diff
@@ -0,0 +1,260 @@
...
+};
+
+type ChoiceRuleAnd = {
+  And: ChoiceRuleComparison[];
+  Next: string;
+};
+
+type ChoiceRuleOr = {
+  Or: ChoiceRuleComparison[];
+  Next: string;
```
Comment (ebisbe) [inline lib/index.d.ts:136]:
```suggestion
  Next: string;
  Comment?: string;
```

File: lib/index.d.ts [hunk 14]
```diff
@@ -0,0 +1,260 @@
...
+};
+
+type ChoiceRuleNot = {
+  Not: ChoiceRuleComparison;
+  Next: string;
+};
+
+type ChoiceRuleAnd = {
+  And: ChoiceRuleComparison[];
+  Next: string;
```
Comment (ebisbe) [inline lib/index.d.ts:131]:
```suggestion
  Next: string;
  Comment?: string;
```

File: lib/index.d.ts [hunk 15]
```diff
@@ -0,0 +1,260 @@
...
+  TimestampGreaterThanEqualsPath?: string;
+  TimestampLessThan?: string;
+  TimestampLessThanPath?: string;
+  TimestampLessThanEquals?: string;
+  TimestampLessThanEqualsPath?: string;
+};
+
+type ChoiceRuleNot = {
+  Not: ChoiceRuleComparison;
+  Next: string;
```
Comment (ebisbe) [inline lib/index.d.ts:126]:
```suggestion
  Next: string;
  Comment?: string;
```

**Comment by zirkelc** [comment]
@horike37 could you please take a look if you have the chance. If you prefer not to include TS types in this package, we can move this PR to the [DefinitelyTyped](https://github.com/DefinitelyTyped/DefinitelyTyped) repo.

File: lib/index.d.ts [hunk 16]
```diff
@@ -0,0 +1,264 @@
...
+  | TaskParametersForSNS
+  | { [key: string]: string | { [key: string]: string } };
+}
+
+interface Pass extends StateBase {
+  Type: 'Pass';
+  Parameters?: {
+    [key: string]: string | Array<unknown> | { [key: string]: string };
+  };
+  Result?: {
```
Comment (ebisbe) [inline lib/index.d.ts:219]:
```suggestion
  Result?: string | {
```
```suggestion
  Result?: string | [] | {
```
I'm using `Result: "${env:something}"` to add an array. Not sure what's the best approach here. The string value resolves into an array. The tip from the Ui console says: `Must be valid JSON.` Array and object are the only ones valid.

**Comment by horike37** [comment]
@zirkelc 
Thank you for this PR! Adding the type system should be really important. But right now we can't regularly maintain this repo, so it would be great if you can move this PR to the [DefinitelyTyped](https://github.com/DefinitelyTyped/DefinitelyTyped) repo.

File: lib/index.d.ts (hunk 16 above)
Comment (zirkelc) [reply lib/index.d.ts:219]: Should we try something like this: https://github.com/microsoft/TypeScript/issues/1897#issuecomment-331765301

```ts
export type JSONPrimitive = string | number | boolean | null;
export type JSONValue = JSONPrimitive | JSONObject | JSONArray;
export type JSONObject = { [member: string]: JSONValue };
export interface JSONArray extends Array<JSONValue> {}
```

```suggestion
  Result?: JSONValue;
```

**Comment by zirkelc** [comment]
@horike37 thanks for you reply! Alright, I'll submit a PR to DefinitelyTyped in the next few days.

**Comment by zirkelc** [comment]
@ebisbe I created a new PR in the DefinetlyTyped repo: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/66693 

I didn't try it yet, but I think to get the types merged from this package, we have to use module augmentation in our `serverless.ts` file directly:

```
import type { AWS } from '@serverless/typescript';
import type { StepFunctions } from 'serverless-step-functions';

declare module '@serverless/typescript' {
   interface AWS {
     stepFunctions?: StepFunctions;
  }
}
```

**Comment by ebisbe** [comment]
> we have to use module augmentation

I don't know what's that it. I will try to help you on the other PR.

**Comment by zirkelc** [comment]
It means add types from package `serverless-step-functions` into the base types of `@serverless/typescript`. However, I figured out that type packages can install other type packages. So we get it to work without add this part at the top:

```
import type { AWS } from '@serverless/typescript';
import type { StepFunctions } from 'serverless-step-functions';

declare module '@serverless/typescript' {
   interface AWS {
     stepFunctions?: StepFunctions;
  }
}
```

We have to wait for the approval on the DT repo.

**Comment by ebisbe** [comment]
Ok. What I don't like on this PR is that we have been doing it manually... It would be perfect if we could extract the full state machine definition and convert it to types so it can be automated. Here it's for CF resources https://github.com/awboost/cfntypes ( only sadly )

**Comment by ebisbe** [comment]
@zirkelc this package has a validation option. Let me see what is using under the hood.

**Pull request closed without merging**